smslibrary/src/main/java/com/eis/smslibrary/SMSPeer.java

```diff
@@ -0,0 +1,123 @@
+package com.eis.smslibrary;
+
+import androidx.annotation.NonNull;
+
+import com.eis.communication.Peer;
+
+/**
+ * Implementation of {@link Peer} for SMS communication channel
+ *
+ * @author Luca Crema, Marco Mariotto
+ */
+public class SMSPeer implements Peer<String>, Comparable<SMSPeer> {
+
+    public static final int MAX_TELEPHONE_NUMBER_LENGTH = 20;
+    public static final int MIN_TELEPHONE_NUMBER_LENGTH = 3;
+
+    private String telephoneNumber;
+
+    /**
+     * @param telephoneNumber a valid telephone number (checkTelephoneNumber state must be TELEPHONE_NUMBER_VALID)
+     */
+    public SMSPeer(String telephoneNumber) {
+        this.telephoneNumber = telephoneNumber;
+    }
+
+    /**
+     * Checks if the given phone number is valid.
+     *
+     * @param telephoneNumber the phone number to check.
+     * @return The {@link TelephoneNumberState} of the telephone number after the tests.
+     */
+    public static TelephoneNumberState checkPhoneNumber(String telephoneNumber) {
+        //Check if the number is shorter than the MAX.
+        if (telephoneNumber.length() > MAX_TELEPHONE_NUMBER_LENGTH) {
+            return TelephoneNumberState.TELEPHONE_NUMBER_TOO_LONG;
+        }
+        //Check if the number is longer than the MIN.
+        if (telephoneNumber.length() < MIN_TELEPHONE_NUMBER_LENGTH) {
+            return TelephoneNumberState.TELEPHONE_NUMBER_TOO_SHORT;
+        }
+        //Check if it's actually a number and doesn't contain anything else
+        //First we have to remove the "+"
+        if (!telephoneNumber.substring(1, telephoneNumber.length() - 1).matches("[0-9]+")) {
+            return TelephoneNumberState.TELEPHONE_NUMBER_NOT_A_NUMBER;
+        }
+        //Check if there is a country code.
+        if (telephoneNumber.charAt(0) != '+') {
+            return TelephoneNumberState.TELEPHONE_NUMBER_NO_COUNTRY_CODE;
+        }
+        //If it passed all the tests we are sure the number is valid.
+        return TelephoneNumberState.TELEPHONE_NUMBER_VALID;
+    }
```

Comment:
Alright, so:

1. As you said, I don't feel the need to have the function static.
2. when you check if there's a '+' in the String, you're returning a number state when there's no "+", so if I give a valid number but with no prefix you won't say it's valid, even though it might be.
3. Is it better to use an enum as an error checking strategy? Aren't you using enums to do the job of exceptions? You can just create custom exceptions using a class, while here you return me a enum (aka a number) that I have to understand and use to know if my number is valid. So I think using exceptions would just be better/faster. You also might just need one exception: `IllegalArgumentException` if the given String is NOT only numbers (with a "+" at the start that might or might not be there), which is pretty easy to do using regex.

Reply:
1) I do feel the need to have the function static. It's way faster to have something like
'SMSPeer.checkPhoneNumber(tel)'
rather than
'new SMSPeer(tel).checkPhoneNumber()'
especially if you are interacting with the user like in inputs.

2) Then what? We can force the library user to put country code in the number, can be done by the app user or by the app programmer (somehow automatically). It's clearer to have the full phone address to avoid uncertainty.

3) Why would you throw an exception on a control? It's not as if the user has instantiated a new peer with an illegal number, it's not an error, it's just a check, and as a check it can have different fails but it's not an error. If you want we could expand constructor specifications so that if the phone number is not `== TelephoneNumberState.TELEPHONE_NUMBER_VALID` it throws an illegalArgumentException that contains the TelephoneNumberState (the failure reason)

Reply:
Alright, I agree on number 2 then, but number 3 stands only if number 1 is true (since on number 3 you say "why would you throw an exception on a control").
I'd suggest this (and I want your opinion):
Remove the static function
Automatically check if the number is valid on object construction (and throw if invali).
This way you get from `new SMSPeer(phone).isValid();` to just `new SMSPeer(phone)` which doesn't throw if thr number is valid.
Based on your first sentence on number 3 the static function becomes useless since it gets transformed from a control to an auto-validator/invalidator.
How does this sound?

Reply:
I still think a static checking method would be way more comfortable when interacting with the user.
Is it really useful something like
`try{
new SMSPeer(number)
catch(IllegalArgumentException e){
  switch(e.????){
 case TELEPHONE_NUMBER_TOO_LONG:
Toast.makeText(ctx,"Number too long!");
break;
}
}`
(Just came to my mind that we would proably need a custom exception so that we can pass the telephoneNumberState)

Wouldn't it be way more reasonable having
` TelephoneNumberState tState = SMSPeer.checkTelephoneNumber(number);
if(tState != VALID_TELEPHONE_NUMBER){
switch(tState){
case TELEPHONE_NUMBER_TOO_LONG:
Toast.makeText(ctx,"Number too long!");
break;
}
}
`

Reply:
Well, if you really want custom error messages than you're totally right (even though it just becomes this:)
```
try{ peer = new SMSPeer(number); }
catch(PhoneNumberTooLong e){
    Toast.makeText(ctx, "Number too long!");
}
catch(InvalidCharacters e){
    Toast.makeText(ctx, "Only numbers are valid!");
}
...
```
I was starting from the assumption that you don't need many different errors if the number is not valid.
But yeah, we can totally fo it your way if you really need different "error states".
(Again, i'd prefer this way ^ since they're still errors, more than states. Like what if the SMSPeer is constructed and the message is sent anyway even though it's invalid, just because the user never checked?)

Reply:
I agree with the fact that we should throw an exception on the constructor, we could use a single exception with the state as a variable so that one doesn't have to write 6 catch but can use a switch, I'll sort this out.

Reply:
Removed non-static telephone number check method and added the exception on the constructor